src/back/hlsl/mod.rs

```diff
@@ -6,29 +6,99 @@ Backend for [HLSL][hlsl] (High-Level Shading Language).
 - 5.1
 - 6.0
 
-# General Matrix Note
+# Layout of values in `uniform` buffers
 
-All matrix construction/deconstruction is row based in HLSL. This means that when
-we construct a matrix from column vectors, our matrix will be implicitly transposed.
-The inverse transposition happens when we call `[0]` to get the zeroth column vector.
+WGSL's ["Internal Layout of Values"][ilov] rules specify how each WGSL
+type should be stored in `uniform` and `storage` buffers. The HLSL we
+generate must access values in that form, even when it is not what
+HLSL would use normally.
 
-Because all of our matrices are implicitly transposed, we flip arguments to `mul`. `mat * vec`
-becomes `vec * mat`, etc. This acts as the inverse transpose making the results identical.
+The rules described here only apply to WGSL `uniform` variables. WGSL
+`storage` buffers are translated as HLSL `ByteAddressBuffers`, for
+which we generate `Load` and `Store` method calls with explicit byte
+offsets. WGSL pipeline inputs must be scalars or vectors; they cannot
+be matrices, which is where the interesting problems arise.
 
-The only time we don't get this implicit transposition is when reading matrices from Uniforms/Push Constants.
-To deal with this, we add `row_major` to all declarations of matrices in Uniforms/Push Constants.
+## Row- and column-major ordering for matrices
 
-Finally because all of our matrices are transposed, if you use `mat3x4`, it'll become `float3x4` in HLSL
-(HLSL has inverted col/row notation).
+WGSL specifies that matrices in uniform buffers are stored in
+column-major order. This matches HLSL's default, so one might expect
+things to be straightforward. Unfortunately, WGSL and HLSL disagree on
+what indexing a matrix means: in WGSL, `m[i]` retrieves the `i`'th
+*column* of `m`, whereas in HLSL it retrieves the `i`'th *row*. We
+want to avoid translating `m[i]` into some complicated reassembly of a
+vector from individually fetched components, so this is a problem.
 
-# Matrix struct member of the form `matCx2` Note
+However, with a bit of trickery, it is possible to use HLSL's `m[i]`
+as the translation of WGSL's `m[i]`:
 
-Struct member matrices of the form `matCx2` are translated to a sequence of C `vec2`s due to
-differences in alignment between WGSL and HLSL for uniform buffers.
+- We declare all matrices in uniform buffers in HLSL with the
+  `row_major` qualifier, and transpose the row and column counts: a
+  WGSL `mat3x4<f32>`, say, becomes an HLSL `row_major float4x3`. Since
```

Comment:
A WGSL `mat3x4<f32>` becomes an HLSL `row_major float3x4` (not `float4x3`) due to HLSL's matrix notation being floatRxC (i.e. inverted relative to WGSL's).

Reply:
Good catch, thank you!